File: algorithms/python/huffman.py
import sys
import time
import heapq

def huffman(text):
    freq = {}
    for c in text: freq[c] = freq.get(c, 0) + 1

    heap  = []
    counter = 0
    for ch, f in freq.items():
        heapq.heappush(heap, (f, counter, ch, None, None))
        counter += 1

    while len(heap) > 1:
        f1, _, c1, l1, r1 = heapq.heappop(heap)
        f2, _, c2, l2, r2 = heapq.heappop(heap)
        heapq.heappush(heap, (f1+f2, counter, None, (f1,c1,l1,r1), (f2,c2,l2,r2)))
        counter += 1

    codes = {}
    def build(node, depth):
        f, _, ch, left, right = node
        if left is None and right is None:
            codes[ch] = depth if depth > 0 else 1
            return
        build(left,  depth+1)
        build(right, depth+1)

    if heap:
        build(heap[0], 0)

    total_bits = sum(freq[c] * codes.get(c, 1) for c in freq)
    return len(freq), total_bits

def main():
    text = sys.stdin.read().strip()
    n    = len(text)

    start          = time.perf_counter()
    unique, total  = huffman(text)
    ms             = (time.perf_counter() - start) * 1000

    ratio = total / (8.0 * n)
    sys.stdout.write(f"{unique}\n{total}\n{ratio:.2f}\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()

File: algorithms/rust/huffman.rs
use std::io::{self, Read, Write, BufWriter};
use std::time::Instant;
use std::collections::BinaryHeap;
use std::cmp::Reverse;

#[derive(Eq, PartialEq)]
enum Node {
    Leaf { freq: usize, ch: u8 },
    Internal { freq: usize, left: Box<Node>, right: Box<Node> },
}

impl Node {
    fn freq(&self) -> usize {
        match self { Node::Leaf{freq,..} | Node::Internal{freq,..} => *freq }
    }
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.freq().cmp(&other.freq())
    }
}
impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

fn build_codes(node: &Node, depth: usize, codes: &mut [usize; 256]) {
    match node {
        Node::Leaf { ch, .. } => {
            codes[*ch as usize] = if depth > 0 { depth } else { 1 };
        }
        Node::Internal { left, right, .. } => {
            build_codes(left,  depth + 1, codes);
            build_codes(right, depth + 1, codes);
        }
    }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let text = input.trim().as_bytes();
    let len  = text.len();

    let mut freq = [0usize; 256];
    for &b in text { freq[b as usize] += 1; }

    let start = Instant::now();

    let mut heap: BinaryHeap<Reverse<Box<Node>>> = BinaryHeap::new();
    let mut unique = 0;
    for (ch, &f) in freq.iter().enumerate() {
        if f > 0 {
            heap.push(Reverse(Box::new(Node::Leaf { freq: f, ch: ch as u8 })));
            unique += 1;
        }
    }

    while heap.len() > 1 {
        let Reverse(l) = heap.pop().unwrap();
        let Reverse(r) = heap.pop().unwrap();
        let f = l.freq() + r.freq();
        heap.push(Reverse(Box::new(Node::Internal {
            freq: f, left: l, right: r
        })));
    }

    let mut codes = [0usize; 256];
    if let Some(Reverse(root)) = heap.pop() {
        build_codes(&root, 0, &mut codes);
    }

    let total_bits: usize = (0..256)
        .filter(|&i| freq[i] > 0)
        .map(|i| freq[i] * codes[i])
        .sum();

    let ms    = start.elapsed().as_secs_f64() * 1000.0;
    let ratio = total_bits as f64 / (8.0 * len as f64);

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    writeln!(out, "{}", unique).unwrap();
    writeln!(out, "{}", total_bits).unwrap();
    writeln!(out, "{:.2}", ratio).unwrap();
    eprintln!("{:.3}", ms);
}